### PR TITLE
Audio library

### PR DESCRIPTION
**Pull request by kattalainen**
Added audio engine and audio class and testing in sandbox main

### CODE REVIEW

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+Audio::Audio()
+{
+    pSourceVoice = nullptr;
+    m_Wfx = { 0 };
+    m_Buffer = { 0 };
+    m_Hr = NULL;
+}
+
+Audio::~Audio()
+{
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp:13]: Minne från medlemsvariabler som är pekare måste frigöras

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+HRESULT Audio::findChunk(HANDLE hFile, DWORD fourcc, DWORD& dwChunkSize, DWORD& dwChunkDataPosition)
+{
+    HRESULT hr = S_OK;
+    if (INVALID_SET_FILE_POINTER == SetFilePointer(hFile, 0, NULL, FILE_BEGIN))
+        return HRESULT_FROM_WIN32(GetLastError());
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp:23]: If-satser ska skrivas i block, enligt vår kodstandard, ex:
if (true)
{
   doSomething();
}

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    DWORD dwRIFFDataSize = 0;
+    DWORD dwFileType;
+    DWORD bytesRead = 0;
+    DWORD dwOffset = 0;
+
+    while (hr == S_OK)
+    {
+        DWORD dwRead;
+        if (0 == ReadFile(hFile, &dwChunkType, sizeof(DWORD), &dwRead, NULL))
+            hr = HRESULT_FROM_WIN32(GetLastError());
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp:38]: If-sats i block

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+        if (0 == ReadFile(hFile, &dwChunkDataSize, sizeof(DWORD), &dwRead, NULL))
+            hr = HRESULT_FROM_WIN32(GetLastError());
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp:43]: Block

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+
+        switch (dwChunkType)
+        {
+        case fourccRIFF:
+            dwRIFFDataSize = dwChunkDataSize;
+            dwChunkDataSize = 4;
+            if (0 == ReadFile(hFile, &dwFileType, sizeof(DWORD), &dwRead, NULL))
+                hr = HRESULT_FROM_WIN32(GetLastError());
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp:53]: Block

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            break;
+
+        default:
+            if (INVALID_SET_FILE_POINTER == SetFilePointer(hFile, dwChunkDataSize, NULL, FILE_CURRENT))
+                return HRESULT_FROM_WIN32(GetLastError());
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp:60]: Block

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        if (dwChunkType == fourcc)
+        {
+            dwChunkSize = dwChunkDataSize;
+            dwChunkDataPosition = dwOffset;
+            return S_OK;
+        }
+
+        dwOffset += dwChunkDataSize;
+
+        if (bytesRead >= dwRIFFDataSize) return S_FALSE;
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp]: Block

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    }
+
+    return S_OK;
+}
+
+HRESULT Audio::readChunkData(HANDLE hFile, void* buffer, DWORD buffersize, DWORD bufferoffset)
+{
+    HRESULT hr = S_OK;
+    if (INVALID_SET_FILE_POINTER == SetFilePointer(hFile, bufferoffset, NULL, FILE_BEGIN))
+        return HRESULT_FROM_WIN32(GetLastError());
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp:89]: Block

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    DWORD dwRead;
+    if (0 == ReadFile(hFile, buffer, buffersize, &dwRead, NULL))
+        hr = HRESULT_FROM_WIN32(GetLastError());
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp:95]: Block

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        strFileName,
+        GENERIC_READ,
+        FILE_SHARE_READ,
+        NULL,
+        OPEN_EXISTING,
+        0,
+        NULL);
+
+    if (INVALID_HANDLE_VALUE == hFile)
+        Log::Print("Invalid handle value\n");
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp:117]: Block

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+    if (INVALID_SET_FILE_POINTER == SetFilePointer(hFile, 0, NULL, FILE_BEGIN))
+        Log::Print("Invalid set file pointer\n");
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.cpp:122]: Block

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.h [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+public:
+	Audio();
+	~Audio();
+
+	void OpenFile(IXAudio2* pXAudio2, const TCHAR* sound);
+	void PlayAudio();
+	void StopAudio();
+
+private:
+	IXAudio2SourceVoice* pSourceVoice;
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.h]: Borde heta m_pSourceVoice, enligt vår kodstandard

File: StortSpelProjekt/Engine/src/AudioEngine/AudioEngine.cpp [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+	if (FAILED(hr = XAudio2Create(&m_pXAudio2, 0, XAUDIO2_DEFAULT_PROCESSOR)))
+	{
+		Log::Print("Failed to create audio engine\n");
+	}
+
+	if (FAILED(hr = m_pXAudio2->CreateMasteringVoice(&m_pMasterVoice)))
+	{
+		Log::Print("Failed to create mastering voice\n");
+	}
+	//Log::Print("CREATED SOME SHET\n");
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/AudioEngine/AudioEngine.cpp]: Städa gärna undan gamla kommentarer

File: StortSpelProjekt/Engine/src/AudioEngine/AudioEngine.cpp [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+AudioEngine::~AudioEngine()
+{
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Engine/src/AudioEngine/AudioEngine.cpp:21]: Måste frigöra minne från pekare

File: StortSpelProjekt/Sandbox/src/main.cpp [hunk 15]
```diff
@@ -191,6 +191,13 @@ int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE, LPSTR, int nCmdShow)
 #pragma endregion CreateScene1
 	char sceneName[10] = "scene0";
 	sceneManager->SetSceneToDraw(sceneManager->GetScene(sceneName));
+
+
+    // AUDIO TESTING, press SPACE to play, TAB to stop
+    Audio audiotest;
+    audiotest.OpenFile(engine.GetAudioEngine(), TEXT("../Vendor/Resources/Audio/AGameWithNoName.wav"));
```
Comment (LeoWikstrom) [inline StortSpelProjekt/Sandbox/src/main.cpp]: Bra val av musik

**Review by LeoWikstrom — CHANGES_REQUESTED** [review]

**Review by LeoWikstrom — DISMISSED** [review]

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.h [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+#include "AudioEngine.h"
+//#include <xaudio2.h>
+//#define XAUDIO2_HELPER_FUNCTIONS
+
+#define fourccRIFF 'FFIR'
+#define fourccDATA 'atad'
+#define fourccFMT ' tmf'
+#define fourccWAVE 'EVAW'
+#define fourccXWMA 'AMWX'
+#define fourccDPDS 'sdpd'
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.h:15]: Add comment to what this is.

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.h [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+public:
+	Audio();
+	~Audio();
+
+	void OpenFile(IXAudio2* pXAudio2, const TCHAR* sound);
+	void PlayAudio();
+	void StopAudio();
+
+private:
+	IXAudio2SourceVoice* m_pSourceVoice;
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.h:32]: What is a audiosourcevoice

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.h [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+	WAVEFORMATEXTENSIBLE m_Wfx;
+	XAUDIO2_BUFFER m_Buffer;
+	HRESULT m_Hr;
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.h]: If these are relevant to the engineSystem, explain what they are.

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.h [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+
+	HRESULT findChunk(HANDLE hFile, DWORD fourcc, DWORD& dwChunkSize, DWORD& dwChunkDataPosition);
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.h:42]: Brief explination in the .h file what it does

File: StortSpelProjekt/Engine/src/AudioEngine/Audio.h [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+	HRESULT readChunkData(HANDLE hFile, void* buffer, DWORD buffersize, DWORD bufferoffset);
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/AudioEngine/Audio.h:44]: Brief explination in the .h file what it does

File: StortSpelProjekt/Engine/src/AudioEngine/AudioEngine.cpp [hunk 21]
```diff
@@ -0,0 +1,29 @@
+#include "AudioEngine.h"
+
+AudioEngine::AudioEngine()
+{
+	m_pXAudio2 = nullptr;
+	m_pMasterVoice = nullptr;
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/AudioEngine/AudioEngine.cpp:6]: What is different with master voice.

File: StortSpelProjekt/Engine/src/AudioEngine/AudioEngine.cpp [hunk 22]
```diff
@@ -0,0 +1,29 @@
+#include "AudioEngine.h"
+
+AudioEngine::AudioEngine()
+{
+	m_pXAudio2 = nullptr;
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/AudioEngine/AudioEngine.cpp:5]: My understanding is that audio2 is the "heart" of the audio system?

File: StortSpelProjekt/Engine/src/AudioEngine/AudioEngine.h [hunk 23]
```diff
@@ -0,0 +1,18 @@
...
+class AudioEngine
+{
+public:
+	AudioEngine();
+	~AudioEngine();
+
+	IXAudio2* GetAudioEngine();
+
+private:
+	IXAudio2* m_pXAudio2;
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/AudioEngine/AudioEngine.h:15]: "heart" of the audio play system?

File: StortSpelProjekt/Engine/src/AudioEngine/AudioEngine.h [hunk 24]
```diff
@@ -0,0 +1,18 @@
...
+	IXAudio2MasteringVoice* m_pMasterVoice;
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/AudioEngine/AudioEngine.h:17]: Master voice handles other voices? How is this used.

File: StortSpelProjekt/Engine/src/Engine.h [hunk 25]
```diff
@@ -42,6 +42,7 @@ class Engine
 
 	Renderer* const GetRenderer() const;
 	SceneManager* const GetSceneHandler() const;
+	IXAudio2* const GetAudioEngine() const;
```
Comment (fjz345) [inline StortSpelProjekt/Engine/src/Engine.h]: Return a AudioEngine* instead of a IXAudio2*.

File: StortSpelProjekt/Sandbox/src/main.cpp (hunk 15 above)
Comment (fjz345) [inline StortSpelProjekt/Sandbox/src/main.cpp]: Testing is fine. engine.GetAudioEngine() should not be in main.cpp, instead create something like AudioEngine.CreateSound(string path) which returns a Audio.

**Review by fjz345 — COMMENTED** [review]
Change commented parts :)

**Review by LeoWikstrom — DISMISSED** [review]

File: StortSpelProjekt/Sandbox/src/main.cpp [hunk 26]
```diff
@@ -184,6 +184,12 @@ int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE, LPSTR, int nCmdShow)
 	char sceneName[10] = "scene0";
 	sceneManager->SetSceneToDraw(sceneManager->GetScene(sceneName));
 
+
+    // AUDIO TESTING, press SPACE to play, TAB to stop. Only for testing purposes, these calls will later be changed to work with the ecs/scene system
+    Audio audiotest;
+    audiotest.OpenFile(engine.GetAudioEngine()->GetAudioEngine(), TEXT("../Vendor/Resources/Audio/AGameWithNoName.wav"));
```
Comment (fjz345) [inline StortSpelProjekt/Sandbox/src/main.cpp]: Just remove this test and we can approve again.

**Review by LeoWikstrom — APPROVED** [review]

**Review by fjz345 — APPROVED** [review]

**Pull request merged**